File: frontend/src/lib/error-log.ts
import { apiFetch } from "@/lib/api";

/** Owner-facing view onto backend/error_alerts.py's basic error log
 * (2026-09-09) — every genuinely unhandled backend error (a crash, not
 * a routine 4xx), most-recent-first. The alert-email address that
 * decides whether one of these also triggers an email lives in
 * `lib/notifications.ts`'s `NotificationSettings.alert_email`. */
export type ErrorLogEntry = {
  timestamp: string;
  method: string;
  path: string;
  error: string;
  traceback: string;
};

export function listErrorLog(limit = 50) {
  return apiFetch<ErrorLogEntry[]>(`/api/agent/error-log?limit=${limit}`);
}

/** Deliberately triggers a real unhandled exception so it flows through
 * the ACTUAL error_alerts.py pipeline (log + cooldown-gated email) —
 * always rejects (the backend responds 500 on purpose), so a caller
 * should treat any thrown ApiError here as "the test ran," not a real
 * failure, then re-check the log/inbox. */
export function triggerTestError() {
  return apiFetch<void>("/api/agent/error-log/test", { method: "POST" });
}

File: backend/apis/error_log.py
"""Owner-facing view onto `error_alerts.py`'s basic error log (2026-09-09)
— a small, dedicated file, same "one concern, one file" precedent as
`apis/chat_settings.py`/`apis/crm_resume.py`. The alert-email address
itself lives in `apis/notifications.py`'s `NotificationSettings`
(alongside Mailgun/Twilio — it's one more "where do alerts go" setting
in that same panel), not here — this file is purely: read the log, and
deliberately trigger the real error path once to prove the whole
log+alert pipeline actually works."""

from fastapi import APIRouter, Depends
from pydantic import BaseModel

from apis.deps import Role, require_role
from error_alerts import read_recent_errors

router = APIRouter(prefix="/agent", dependencies=[Depends(require_role(Role.admin, Role.owner))])


class ErrorLogEntry(BaseModel):
    timestamp: str
    method: str
    path: str
    error: str
    traceback: str


@router.get("/error-log", response_model=list[ErrorLogEntry])
def get_error_log(limit: int = 50) -> list[ErrorLogEntry]:
    """Most-recent-first. Reads directly from the JSONL file — no DB
    table for this (see error_alerts.py's own docstring for why "basic"
    was the right scope here)."""
    return [ErrorLogEntry(**entry) for entry in read_recent_errors(limit=limit)]


@router.post("/error-log/test")
def trigger_test_error() -> None:
    """Deliberately raises an unhandled exception so it flows through the
    REAL path (main.py's global exception handler -> error_alerts.py's
    logging + cooldown-gated email alert) — a genuine end-to-end test of
    the whole pipeline, not a separate, parallel "pretend send" like the
    notification panel's own test-email/test-sms buttons. The owner sees
    a plain 500 in the dashboard (expected) and should then check the
    error log / their alert inbox."""
    raise RuntimeError("This is a test error, deliberately triggered from the dashboard's error-log panel.")
